docs/contracts/transient-storage.rst

```diff
@@ -0,0 +1,61 @@
+.. index:: ! transient storage
+
+.. _transient-storage:
+
+*****************
+Transient Storage
+*****************
+
+Transient storage is another data location besides memory, storage, calldata
+(and return-data and code) which was introduced alongside its respective opcodes
+``TSTORE`` and ``TLOAD`` by `EIP-1153 <https://eips.ethereum.org/EIPS/eip-1153>`_.
+This new data location behaves as a key-value store similar to storage with the main
+difference being that data in transient storage is not permanent, but is scoped to
+the current transaction only, after which it will be reset to zero. The values on
+transient storage are never deserialized from storage or serialized to storage,
+thus the gas cost of operating in transient storage is much cheaper,
+since it doesn't require disk access.
+
+Currently, the compiler can parse ``transient`` as a data location, however it is not
+defined as a keyword of the language yet. This means that the use of ``transient``
+is backwards-compatible and does not break previous code that eventually used it as an identifier.
+Note however that such use of ``transient`` as a data location is only allowed for
+:ref:`value type <value-types>` state variable declarations. Further support for
+other types as well as code generation is intended with upcoming releases.
```

Comment:
I'd remove any promises from this. This is our plan, but it does not necessarily have to be encoded in the docs. We already said that in release notes and IMO that was a better place for a statement of this kind.

I'd instead just plainly describe what isn't supported: reference types, mappings, references to transient storage, locals, parameters, etc.